c/enc/hash_forgetful_chain_inc.h

```diff
@@ -210,79 +212,124 @@ static BROTLI_INLINE void FN(FindLongestMatch)(
                                                     &data[cur_ix_masked],
                                                     max_length);
-        if (len >= 4) {
-          /* Comparing for >= 3 does not change the semantics, but just saves
-             for a few unnecessary binary logarithms in backward reference
-             score, since we are not interested in such short matches. */
-          score_t score = BackwardReferenceScore(len, backward);
+        /* If have a backward_references array cut the copy_len so
+           it won't intersect with any other stored references */
+        if (back_refs_size > 0 && *back_refs_position < back_refs_size &&
+            cur_ix + len > backward_references[*back_refs_position].position) {
+          len = backward_references[*back_refs_position].position - cur_ix;
+        }
+        if (len >= 2) {
+          score_t score = BackwardReferenceScoreUsingLastDistance(len);
           if (best_score < score) {
-            best_score = score;
-            best_len = len;
-            out->len = best_len;
-            out->distance = backward;
-            out->score = best_score;
+            if (i != 0) score -= BackwardReferencePenaltyUsingLastDistance(i);
+            if (best_score < score) {
+              best_score = score;
+              best_len = len;
+              out->len = best_len;
+              out->distance = backward;
+              out->score = best_score;
+            }
           }
         }
       }
     }
-    FN(Store)(self, data, ring_buffer_mask, cur_ix);
+    {
+      const size_t bank = key & (NUM_BANKS - 1);
+      size_t backward = 0;
+      size_t hops = self->max_hops;
+      size_t delta = cur_ix - addr[key];
+      size_t slot = head[key];
+      while (hops--) {
+        size_t prev_ix;
+        size_t last = slot;
+        backward += delta;
+        if (backward > max_backward || (CAPPED_CHAINS && !delta)) break;
+        prev_ix = (cur_ix - backward) & ring_buffer_mask;
+        slot = banks[bank].slots[last].next;
+        delta = banks[bank].slots[last].delta;
+        if (cur_ix_masked + best_len > ring_buffer_mask ||
+            prev_ix + best_len > ring_buffer_mask ||
+            data[cur_ix_masked + best_len] != data[prev_ix + best_len]) {
+          continue;
+        }
+        {
+          size_t len = FindMatchLengthWithLimit(&data[prev_ix],
+                                                      &data[cur_ix_masked],
+                                                      max_length);
+          /* If have a backward_references array cut the copy_len so
+             it won't intersect with any other stored references */
+          if (back_refs_size > 0 && *back_refs_position < back_refs_size &&
+              cur_ix + len > backward_references[*back_refs_position].position) {
+            len = backward_references[*back_refs_position].position - cur_ix;
+          }
+          if (len >= 4) {
```

Comment:
why "4"?

Reply:
The same as the next comment, not my code

Reply:
4 means we don't want copies of length 1-3 (as  those, most likely, have negative score).